### ISSUE TITLE
Question: Support for StringIO

### ISSUE THREAD

**Issue by josegomezr** [ikapelyukhin/repomd-parser#15]
Are there plans for supporting StringIO? The use case is: not persisting to disk the downloaded XML.

```ruby
response = http_client.get('https://a-webserver.here/repodata/primary.xml.gz')
response.body.class # => string

# Then I'd have to persit it on disk here
File.write('primary.xml.gz')
entries = RepomdParser::PrimaryXmlParser.new('primary.xml.gz').parse
```

But it looks like it doesn't play well in `overlayfs` as we're hitting `DiskPressure` sometimes, it'd be ideal not to have to persist it in disk :thinking:

**Comment by ikapelyukhin** [comment]
The initializer for `BaseParser` class right now takes the filename as an argument. Compression type is determined based on the filename extension (which seems to me a bit fragile). Instead, it would need to take an IO object (e.g., `File` or `StringIO`) and compression type as parameters.

I can implement it if that's what you need. It would break backward compatibility, so that'd be a new major version release then.

**Comment by josegomezr** [comment]
> The initializer for `BaseParser` class right now takes the filename as an argument. Compression type is determined based on the filename extension (which seems to me a bit fragile). Instead, it would need to take an IO object (e.g., `File` or `StringIO`) and compression type as parameters.

Maybe (not to break compat) an additional constructor? :thinking: `...Parser.from_content(content, reader: :raw|:gzip|:zstd)` ? 

and then with that `reader` the potential decompression is handled as needed? :thinking:

**Comment by ikapelyukhin** [comment]
> Maybe (not to break compat) an additional constructor? 🤔 `...Parser.from_content(content, reader: :raw|:gzip|:zstd)` ?

You'd still have to call the constructor from the `from_content` method and the constructor still expects filename as input.

It would be possible to add `Parser.from_filename` for backward compatibility, which would open the file and determine compression type in the same fashion it currently works and then pass those arguments to a constructor.

It would be also possible to have the constructor accept both filenames and IO objects and check if the argument is inherited from the `IO` class, but having just one type of input parameters would be cleaner.

**Comment by josegomezr** [comment]
Hi @ikapelyukhin, sorry for the delay.

I dug a bit deeper and what you suggested:

> It would be possible to add `Parser.from_filename` for backward compatibility,
> which would open the file and determine compression type in the same fashion
> it currently works and then pass those arguments to a constructor.

Sounds very reasonable so the changes are minimized when updating the gem in rmt.

On the other side, I got a MVP working with the gem, but with the xml file contents coming from faraday:

```ruby
@packages ||= begin
  parser = RepomdParser::PrimaryXmlParser.new(nil)
  Nokogiri::XML::SAX::Parser.new(parser).parse(file_contents)
  @file_contents = nil # reclaim some space

  # PENDING: This part is quite ugly
  files = parser.instance_variable_get(:@referenced_files).tap do |pkgs|
    logger.info "It contains #{pkgs.count} package descriptions"
  end
end
```

At the very least an `attr_accessor :referenced_files` would totally make it prettier hehe.

From my side I need to improve and learn how to get from Faraday an StringIO of the body :thinking: but with the snippet above memory consumption improved a lot when processing >40k packages primary.xml files

**Comment by ikapelyukhin** [comment]
@josegomezr please see #16.

> At the very least an `attr_accessor :referenced_files` would totally make it prettier hehe.

There's no `attr_accessor` because it isn't really supposed to be public -- normally it is a return value from the `parse` method. In 1.0.0 I've actually made sure that `@referenced_files` is emptied -- otherwise the memory wouldn't be reclaimed if you keep the parser object around and you'd get weird results if you call the `parse` method multiple times.

> From my side I need to improve and learn how to get from Faraday an StringIO of the body 🤔 but with the snippet above memory consumption improved a lot when processing >40k packages primary.xml files

Here's the sample for 1.0.0:

```ruby
#!/usr/bin/env ruby

require 'net/http'
require 'uri'
require 'stringio'
require 'repomd_parser'

def parse_primary_xml(primary_xml_link)
  uri = URI(primary_xml_link)
  response = Net::HTTP.get(uri)

  # StringIO is wrapped with Zlib::GzipReader or RepomdParser::ZstdReader
  # here to handle decompression
  io = RepomdParser.decompress_io(
    StringIO.new(response),
    File.basename(uri.path)
  )

  packages = RepomdParser::PrimaryXmlParser.new.parse(io)
  packages.each do |package|
    puts package.name
  end
end

puts 'primary.xml with gzip compression:'
parse_primary_xml('https://download.opensuse.org/update/leap/15.0/non-oss/repodata/9127f8fa369240d051ec4b9a6d93eccaedfff24887012845cc2ae61f0a507741-primary.xml.gz')

puts
puts 'primary.xml with zstd compression:'
parse_primary_xml('https://download.opensuse.org/tumbleweed/repo/non-oss/repodata/898f6f825a1591d890719cef573ac777c3e3ae682c1055399f41fb385d139bb898151cc972f220dfbccc54ca30bd100044695f097264b2efaa3301674d66d250-primary.xml.zst')
```

**Comment by josegomezr** [comment]
This looks promising, I'll benchmark it and come back to you with the results

**Comment by josegomezr** [comment]
I got metrics now! Measured them with `psrecord`.

I got the memory behavior of my original implementation (extracting the internals of the parser class):

zSTD:
![repomd-zstd](https://github.com/ikapelyukhin/repomd-parser/assets/5963207/46ec0991-4d51-4da9-be51-38c9836cdb10)

GZIP:
![repomd-gz-huge](https://github.com/ikapelyukhin/repomd-parser/assets/5963207/bf151045-c54f-4933-a330-2f58498368f3)

---

Versus v1.0.0:

zSTD:
![repomd-zstd](https://github.com/ikapelyukhin/repomd-parser/assets/5963207/0a06f19d-1755-4ef8-a931-2e9bdc909858)

GZIP:
![repomd-gz-huge](https://github.com/ikapelyukhin/repomd-parser/assets/5963207/9e9f8c6f-734e-436d-b3e1-6e36602850c5)

which align very well with another mad scientist experiment I did by turning the parser into an event emitter :scientist: 
(Rough impl here: https://gist.github.com/josegomezr/e1c1d1e3390bea268b0be45de58d18a4)

I think v1.0.0 is winner here (compared against my not the simplest emitter implementation).

**Comment by ikapelyukhin** [comment]
@josegomezr alright, I've published v1.0.0 to Rubygems. :raised_hands:
